### PR TITLE
feat(accountsdb): Generate snapshots, fix UAF, improve bincode

### PR DESCRIPTION
**Pull request by InKryption**
Note: there are a number of auxiliary changes that were made in the course of this change set as a result of running into and needing to fix issues, mainly involving manual testing, and minor amendments, alongside some small refactors.

This implements the most fundamental building blocks for generating snap shot tar files, and sets up our infrastructure proper for then compressing said tar files into zstd archives. At present none of this is plugged in anywhere, it's just unit tested. As such, this code will still probably need to experience much evolution as we sculpt the architecture of consensus and any other related components.

Edit:
the scope of this PR increased from the initial goal of generating snapshots, to include indirect but auxiliary improvements to lay the infrastructure and fix things that will likely be in the scope of snapshot generation.
Also I fixed a UAF in `validator`.

### CODE REVIEW

**Review by 0xNineteen — CHANGES_REQUESTED** [review]

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:450]: Technically there's a way to validate this kind of thing using type info at comptime, eliding the runtime error, thing is it's a fair bit more complicated and a bit less general.

In this case we can obviously see the error is unreachable, since the arguments could never overflow based on the max args. I can look into doing this anyway if you want, lmk if you think the tradeoff is still worth it.

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2008]: 62dd8443ef0c33d5a0fd97cde72b9e1c8ab6a81c

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:245]: 0401ba721134b59988c96ebe959d71ce6c48e5f1

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:383]: 0401ba721134b59988c96ebe959d71ce6c48e5f1

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:780]: 0401ba721134b59988c96ebe959d71ce6c48e5f1

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2020]: a371856ea8cc96f19e8ab02f045dd5739f85f3f1

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2025]: 7ccce8e9ada95180701c99fd4216dfcda0ac5298

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2040]: 7facf5f15096711dbe669f7c596a80dac32cee20

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2219]: 6a2725e8be63f0e8f47a59ed3ed8ddb179945220

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2217]: 6a2725e8be63f0e8f47a59ed3ed8ddb179945220

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2235]: ac9e2b7d61ace0aa9ee6049b8af466b8a9f65103

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2236]: ac9e2b7d61ace0aa9ee6049b8af466b8a9f65103

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2068]: 238b5cf0ef10f82d6bc784bf03c1e9e276886401

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2066]: a371856ea8cc96f19e8ab02f045dd5739f85f3f1

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2067]: We have access to `@import("zstd").Writer`, which we can pipe this into, we just don't do that anywhere yet.

**Review by 0xNineteen — CHANGES_REQUESTED** [review]

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:445]: rn we use this function for only integers - what if we only support integers which would make it comptime checkable and we could remove the max_args parameter - then if we want to add more type support we can add as we need - wdyt?

File: src/bincode/optional.zig
Comment (0xNineteen) [inline src/bincode/optional.zig:36]: do we need skipWrite if the serializer returns early in the null case?

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:445]: Sounds like a fair compromise - the alternative use case I was thinking of was a situation where we had a variable-length string with a comptime-known upper bound, there doesn't exist a neat way to express that kind of thing only through types.

Although, maybe such a use case could be supported by requiring non-integer types to expose a decl like `pub const max_len: usize = value;`, so then you could wrap the string with the upper bound in a type which exposes that upper bound.

I'll go ahead with the former, lmk if you're for or against the latter.

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2062]: Well the thing is the functionality is already "exposed", you just wrap a writer in the zstd writer, and then pass that to this function, and voila, compression. There just isn't anywhere where this makes sense to do. I can go ahead and throw an example in a demonstrative test.

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:247]: c430c7420fa546786f62843bb7ed001bb708d48

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:279]: c430c7420fa546786f62843bb7ed001bb708d48

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:396]: c430c7420fa546786f62843bb7ed001bb708d48

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2269]: ec98723a357d501d1139786e1320bf9286de43d

File: src/bincode/bincode.zig
Comment (InKryption) [reply src/bincode/bincode.zig:676]: Isn't `hashmap_config` a bit redundant when it's already the field of a config type?

As for the type, it would say it'd probably be a fair bit harder to realize in the particular way you're suggesting:
```suggestion
        hashmap: ?if (hashMapInfo(T)) |hm_info| struct {
            key: FieldConfig(hm_info.Key) = .{},
            value: FieldConfig(hm_info.Value) = .{}
        } else noreturn = null,
```
it needs to be conditional on the hash map type info regardless, because the `FieldConfig` instantiations will need the key and value types.

The other problem is that it can't just be inlined, because we need to be able to pass `HashMapConfig` to the hash map functions.

As a compromise, here's what getting rid of `MaybeHashMapConfig` and inlining its logic would look like:
```suggestion
        hashmap: if (hashMapInfo) |hm_info| HashMapConfig(hm_info) else struct {} = .{},
```

File: src/bincode/optional.zig
Comment (InKryption) [reply src/bincode/optional.zig:36]: That's a fair question. Probably not.
I wrote this after I came up with the skip predicate - maybe it makes the predicate unnecessary, let play around with this.

File: src/bincode/optional.zig
Comment (InKryption) [reply src/bincode/optional.zig:9]: It's mostly muscle memory. The reason I default to this style is because structs which are not meant to be instantiated are supposed to be snake_case, as per the zig style guide, and the name `gen` is just an abbreviation for "generated", ie decls generated based on the comptime parameters/types of the parameters.
I'm fine with renaming these to `S` if you want.

File: src/bincode/bincode.zig
Comment (InKryption) [reply src/bincode/bincode.zig:676]: ebddd6d70c796aae3e406ddab297ce9097bf824

lmk wyt

File: src/bincode/bincode.zig
Comment (InKryption) [reply src/bincode/bincode.zig]: fa420287d16b7bdcb484cd1785f4053fcab52140

File: src/bincode/bincode.zig
Comment (InKryption) [reply src/bincode/bincode.zig]: well actually that breaks `optional.defaultToNulOnEof`, because it also needs the empty type to be compatible.

Comment (InKryption) [reply src/bincode/bincode.zig]: 08ae56850b45b4f1005f9f68837016b306a000a3

File: src/bincode/optional.zig
Comment (InKryption) [reply src/bincode/optional.zig]: 4ca848b798c075df8cde9c567f2cec9e7a1732b4

I dove in and pulled out a bunch of redundancies, this predicate included.

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2182]: fc07181a05877e39764136b050255d08b60cba4b

File: src/accountsdb/snapshots.zig
Comment (InKryption) [reply src/accountsdb/snapshots.zig]: 37edadbabd054f58ce35385a002301aab339f46f

**Comment by InKryption** [comment]
While writing up docs for snapshot generation and putting the semantics to written word, I realized some of the logic actually didn't make sense - I fixed it, realized some obvious improvements, and decided it made sense to bundle them up in this PR.

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig]: yeah probably better to go with the simpler approach for now and add a comment which describes the latter soln for future work if we need it

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:2066]: can you add an other method 'writeSnapshot' which uses the zstd decompression writer by default (probs should be in another pr) - also probs worth thinking about efficient ways to do this - since this blocks flush/clean/... - ideally we can start compression after an account file has been added to the tar archive buffer - would likely need multiple threads ...

File: src/bincode/optional.zig
Comment (0xNineteen) [reply src/bincode/optional.zig:17]: makes sense - would probs be better for the sake of consistency to follow the S pattern

File: src/bincode/bincode.zig
Comment (0xNineteen) [inline src/bincode/bincode.zig:688]: dead code

File: src/accountsdb/db.zig
Comment (0xNineteen) [inline src/accountsdb/db.zig:2032]: would ideally want something more explicit like a incremental bool parameter to create incremental snapshots - or even a separate method explicitly for incremental snapshots - leaning towards latter but will leave it up to you

File: src/accountsdb/db.zig
Comment (0xNineteen) [inline src/accountsdb/db.zig:2043]: one detail about incremental snapshots - they can only be created and paired with an existing full snapshot (something we will need to track) - and the slots which incrementals contain will be (base_slot, highest_rooted_slot] - also when we create a new incremental snapshot we will want to delete older incrementals - and when we create a new full snapshot we will want to delete old full and old incremental snapshots

realizing how this is a fair amount of additions lets leave it to another pr and just revert and keep this pr to only full snapshot creations

**Review by 0xNineteen — COMMENTED** [review]
just some dead code and comments on what we need in future prs - otherwise lgtm after you address the other convos - also should be good to get @dnut approval since its a fairly large pr and we walked through it already

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig]: a26b0119023ec1815e43997fc2f2f90393ae072f

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig]: a26b0119023ec1815e43997fc2f2f90393ae072f

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:2058]: Yeah, that was vaguely how I was picturing it, just like I said, this doesn't slot in anywhere yet so it's hard to judge the best abstraction for it, the main thing I've focused on here is building the infrastructure to make it easy enough to integrate when the time comes.

File: src/bincode/optional.zig
Comment (InKryption) [reply src/bincode/optional.zig]: 50e456e3b1989fcd0ba2797c2cd93e28d4d4384e

File: src/bincode/bincode.zig
Comment (InKryption) [reply src/bincode/bincode.zig]: 50e456e3b1989fcd0ba2797c2cd93e28d4d4384e

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig]: I think the code as is is relatively self-descriptive. Best case scenario is we never actually need to use this for anything other than printing integers - if a more complex use case arises, we can adapt the code as needed without limiting the design space right now.

File: src/accountsdb/snapshots.zig
Comment (0xNineteen) [inline src/accountsdb/snapshots.zig:344]: why not just use the allocator parameter thats passed in to this fcn?

Comment (InKryption) [reply src/accountsdb/snapshots.zig:344]: the point is that it should fail if it tries to allocate, because it should never allocate - we're deserializing an integer.

Comment (InKryption) [reply src/accountsdb/snapshots.zig:344]: There should maybe be a dedicated function for this.

Comment (0xNineteen) [reply src/accountsdb/snapshots.zig:344]: makes sense - yeah dedicated function `readWithoutAlloc` might be a good bincode function addition which just calls `read` with the fb allocator you have here

Comment (InKryption) [reply src/accountsdb/snapshots.zig]: I decided to go with a dedicated function for reading ints - the `readWithoutAlloc` idea is interesting, but it seems hard to verify such a thing in the general case.

File: src/accountsdb/db.zig
Comment (dnut) [inline src/accountsdb/db.zig:279]: Why was this code here in the first place? Is it possible that the map could be out of sync with the filesystem?

File: src/bincode/bincode.zig
Comment (dnut) [inline src/bincode/bincode.zig:674]: This feels oddly specific. Do you foresee adding other fields here for other commonly used types?

My instinct is that HashMap and ArrayList should just use FieldConfig like any other struct that requires custom logic.

The problem with that is you need to specify a field config everywhere you use these types, which is annoying. A compromise would be that they're defined within bincode and used by default for those types, but they are still just a normal FieldConfig, and the custom logic outside FieldConfig is minimal.

I don't expect you to rewrite anything in this PR, just curious about your thoughts on this topic.

File: src/accountsdb/snapshots.zig
Comment (InKryption) [reply src/accountsdb/snapshots.zig:51]: Mainly for order determinism. For one it makes it trivial to verify that the exact data which was read from the bincode is written the same way; secondly, we do a lot of iteration over a lot of the maps in the snapshot, which is something that array hash map is highly optimized for. In this specific case, we don't seem to iterate over `stake_delegations` right now anywhere in the code, but the former point still stands, and we may end up needing to iterate over this at some point in the future whenever we start needing to interact with it.

As for array hash map disadvantages: it doesn't proclaim the same speed in lookups, insertion, and deletion - it doesn't state it's very slow though to be clear, and the zig compiler codebase makes heavy use of it.
And since a large amount of our interactions with these hash maps is iteration, we wouldn't strongly benefit from the normal hash map's presumably enhanced lookup/insertion/deletion

File: src/utils/fmt.zig
Comment (InKryption) [reply src/utils/fmt.zig:6]: Yeah, you deduced it, `@TypeOf(.{ 123 }) == struct { comptime u32 = 123 }`. Comptime fields must always have a default value, and must always be their default value - they're essentially glorified decls, kind of like the `.len` on an array.

Since `comptime_int` doesn't have an actual upper or lower bound, this is the only ergonomic way to get a precise buffer size for it in this case.

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:279]: If the file map of the snapshot doesn't accurately describe the files of the snapshot it's in, that'll become apparent later on in `loadAndVerifyAccountsFiles` if any of the files described by the file map don't exist - if any files beyond what it describes exist, we also don't care anyways, since it has to be included in the file map. This check is just unnecessary work, especially when the filesystem could contain hundreds or thousands of files, and the actual file map could realistically contain a mere fraction of that.

File: src/bincode/bincode.zig
Comment (InKryption) [reply src/bincode/bincode.zig:674]: I have a broader plan to extract this out of FieldConfig as well, which involves actually generalizing `FieldConfig` into `Config`, and providing this functionality through a `hash_map` namespace.
So my thoughts are: yeah this is icky, but it works and helps me further my goal of bincode simplification. Just out of scope for this PR.

File: src/cmd/cmd.zig
Comment (dnut) [reply src/cmd/cmd.zig:422]: makePath seems slightly preferable just for clarity/integrity/atomicity of implementation, but I'm fine leaving it as is.

**Review by dnut — APPROVED** [review]

**Review by 0xNineteen — APPROVED** [review]

**Pull request merged**
